File: 0823-binary-trees-with-factors/0823-binary-trees-with-factors.cpp
class Solution {
public:
    int numFactoredBinaryTrees(vector<int>& arr) {
        int MOD = 1000000007;
        sort(arr.begin(), arr.end()); // Sort the array.

        unordered_map<int, long long> dp; // Use a map to store the number of trees for each value.

        for (int i = 0; i < arr.size(); i++) {
            dp[arr[i]] = 1; // A single node tree with the value arr[i].
            for (int j = 0; j < i; j++) {
                if (arr[i] % arr[j] == 0) { // If arr[j] is a factor of arr[i].
                    int factor = arr[i] / arr[j];
                    if (dp.count(factor)) {
                        dp[arr[i]] = (dp[arr[i]] + dp[arr[j]] * dp[factor]) % MOD;
                    }
                }
            }
        }

        long long result = 0;
        for (auto& entry : dp) {
            result = (result + entry.second) % MOD;
        }

        return static_cast<int>(result);
    }
};

File: 0823-binary-trees-with-factors/0823-binary-trees-with-factors.py
class Solution:
    def numFactoredBinaryTrees(self, arr: List[int]) -> int:
        MOD = 10 ** 9 + 7
        arr.sort()

        location = {val: i for i, val in enumerate(arr)}
        visited = {}
        def dp(index):
            if index in visited:
                return visited[index]
            value = arr[index]
            ways = 1
            for factor1 in arr[:index]:
                factor2 = value // factor1
                if value % factor1 == 0 and factor2 in location:
                    ways += dp(location[factor1]) * dp(location[factor2])
            visited[index] = ways % MOD
            return visited[index]
        
        res = 0
        for i in range(len(arr)):
            res += dp(i)
        return res % MOD